quic/s2n-quic-core/src/transport/parameters/mod.rs

```diff
@@ -1063,6 +1070,119 @@ impl From<connection::id::LocalId> for InitialSourceConnectionId {
+/// and by the server to indicate which version it is using
+#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd, Eq, Ord)]
+pub struct DcSupportedVersions([VarInt; 4]);
+
+impl DcSupportedVersions {
+    /// Create a `DcSupportedVersions` for the given `supported_versions` the client supports
+    pub fn for_client<I: IntoIterator<Item = u32>>(supported_versions: I) -> Self {
+        let mut value = [VarInt::default(); 4];
+
+        for (index, version) in supported_versions.into_iter().enumerate() {
+            value[index] = VarInt::from_u32(version);
+
+            debug_assert!(index < 4, "Only 4 supported versions are supported");
+            ensure!(index < 4, break);
+        }
+
+        DcSupportedVersions(value)
+    }
+
+    /// Create a `DcSupportedVersions` for the `supported_version` the server has selected
+    pub fn for_server(supported_version: u32) -> Self {
+        DcSupportedVersions([
+            VarInt::from_u32(supported_version),
+            VarInt::default(),
+            VarInt::default(),
+            VarInt::default(),
+        ])
+    }
+
+    /// The version the server has selected
+    ///
+    /// Returns `None` if no version was selected
+    pub fn selected_version(&self) -> Result<Option<u32>, DecoderError> {
+        let mut selected_version = None;
+
+        for version in self.into_iter() {
+            if version > 0 {
+                decoder_invariant!(
+                    selected_version.is_none(),
+                    "multiple versions selected by the server"
+                );
+                selected_version = Some(version);
+            }
+        }
+
+        Ok(selected_version)
+    }
+}
+
+impl TransportParameter for DcSupportedVersions {
+    const ID: TransportParameterId = TransportParameterId::from_u32(0xdc0000);
+    type CodecValue = Self;
+
+    fn from_codec_value(value: Self::CodecValue) -> Self {
+        Self(value.0)
+    }
+
+    fn try_into_codec_value(&self) -> Option<&Self::CodecValue> {
+        if *self == Self::default_value() {
+            None
+        } else {
+            Some(self)
+        }
+    }
+
+    fn default_value() -> Self {
+        Self::default()
+    }
+}
+
+impl EncoderValue for DcSupportedVersions {
+    fn encode<E: Encoder>(&self, buffer: &mut E) {
+        for version in self.0 {
+            version.encode(buffer)
+        }
+    }
+}
+
+decoder_value!(
+    impl<'a> DcSupportedVersions {
+        fn decode(buffer: Buffer) -> Result<Self> {
+            let mut versions = [VarInt::default(); 4];
+
+            let (version, buffer) = buffer.decode::<VarInt>()?;
+            versions[0] = version;
+            let (version, buffer) = buffer.decode::<VarInt>()?;
+            versions[1] = version;
+            let (version, buffer) = buffer.decode::<VarInt>()?;
+            versions[2] = version;
+            let (version, buffer) = buffer.decode::<VarInt>()?;
+            versions[3] = version;
```

Comment:
Same here. We shouldn't always expect there to be 4. We should handle the cases that there's less than as well. I would say we need at least 1 entry to be valid though.

Reply:
I think 0 is valid, it would be similar to `max_idle_timeout` where you can either omit the transport parameter or send a value of 0.